### DISCUSSION TITLE
Get extents of the model currently in view

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#1856] [Q&A] [answered]
Is there a way to get the extents of the model that is currently in displayed in the view after zoom (window) operation

**Comment by pmconne** [comment]
Please clarify.
A spatial view can contain any number of models. The extents of each model is defined by the geometry contained within it, so they do not change when you zoom in.
The [iModel itself](https://www.itwinjs.org/learning/imodels/) has its own [extents](https://www.itwinjs.org/reference/imodeljs-common/imodels/imodel/projectextents/) which likewise do not change when you zoom in.

**Reply by ovaisa** [reply]
Sorry, perhaps my use of the word 'extents' was misleading.

We have multiple use cases that are based on the portion of the model that is currently visible in the viewport.  One of the use case is that after user zooms into an area of the model, we want to give them an option to clip (approx) the portion of the model this is currently visible in the viewport...(I am using code from view-clip-sample)

    const block: ClipShape = ClipShape.createBlock(range, ...);
    const clip: ClipVector = ClipVector.createEmpty();
    clip.appendReference(block);
    vp.view.setViewClip(clip)

I want to find out the 'range' that I can use to set this clip.

If I can get this 'range', I can also use it to solve other use cases that we have that are dependent on the portion of model that is currently visible on the screen.  We only have one model in our iModel.

**Reply by ovaisa** [reply]
Hopefully images below clarify bit better the capability we are trying to provide.  The user zooms into area of interest in the model (left image) and then selects option to clip this area (right image).  We will activate the clipping tool to refine the clipped region.

![ClipExtents](https://user-images.githubusercontent.com/83007711/125618782-844c5dc5-af1a-4bc4-a739-7949a3a5828c.jpg)
.

**Comment by pmconne** [comment, marked as answer]
Thanks for clarifying.

The volume of space viewed by a viewport is defined by its [Frustum](https://www.itwinjs.org/reference/imodeljs-common/views/frustum/) which you can obtain via [Viewport.getFrustum](https://www.itwinjs.org/reference/imodeljs-frontend/views/viewport/getfrustum/). Note that this volume can be a box, or a truncated pyramid - [this article](https://www.itwinjs.org/learning/frontend/views/#viewstate-parameters) has some useful illustrations.
You can create a ConvexClipPlaneSet from a Frustum using its getRangePlanes method.

The full range of a geometric model can be obtained via [IModelConnection.Models.queryModelRanges](https://www.itwinjs.org/reference/imodeljs-frontend/imodelconnection/imodelconnection.models/). You could compute the intersection of the model range with the viewport frustum, but typically you'd end up with an awkward many-sided clip volume.

From your screenshots it looks like you're trying to find the intersection of the viewport's left and right clipping planes with the closest wall of the building, then define a clip box by extruding that plane to the viewport's back clipping plane? You'd need some information beyond model range and viewport frustum for that to work, unless you define "closest wall of the building" as "closest plane of the model's range box".

**Reply by ovaisa** [reply]
Thank you Paul for this explanation.  I did look at Frustrum, seems to work well when camera in on, but will look at it again for parallel projection case (which is what I expect users will be using).  I was hoping I could use some already available function (that I couldn't find). 

The article you pointed to has good info, I should be able to use this to get what I need.  It doesn't have to be exact, something to give user a convenient starting point to work with clipping tool.

I only have one model so hopefully it should keep things simple. 

Thank you for your guidance!

**Reply by ovaisa** [reply]
From the information in the article, using deltaX and deltaY in the image below should give me the approximate view volume I am looking for.

![Frustrum](https://user-images.githubusercontent.com/83007711/125791339-2507f7d7-4a12-4e26-a256-26ef02e5063a.JPG)

The following line in the article states that delta is stored with the view state

The ViewState holds the parameters: {origin}{delta}[Matrix3d] from which the View frustum is derived.

I was unable to find a delta property in any of the view states (there was one in 2d view definition but none in any 3d).  I did find a property named 'extents' which seems to change with the view (becomes small as I zoom in - which is what I would expect to happen to deltaXY).  Can you please confirm if this property is indeed deltaXYZ and if not, how can I get deltaXY.

Thank you.

**Reply by bbastings** [reply]
Yes, ViewState3d.extents is the frustum extents.
